File: design_patterns/builder/builder_java/src/main/java/app/swarch/designpatterns/builder/BuilderDemo.java
package app.swarch.designpatterns.builder;

class Pizza {

    private final Boolean big;
    private final String cheese;
    private final String meat;

    public Pizza(Boolean big, String cheese, String meat) {
        this.big = big;
        this.cheese = cheese;
        this.meat = meat;
    }

    public Boolean getBig() {
        return big;
    }

    public String getCheese() {
        return cheese;
    }

    public String getMeat() {
        return meat;
    }

    @Override
    public String toString() {
        return "Pizza{" +
            "big=" + big +
            ", cheese='" + cheese + '\'' +
            ", meat='" + meat + '\'' +
            '}';
    }
}

class Receipt {

    private final Boolean big;
    private final String cheese;
    private final String meat;

    public Receipt(Boolean big, String cheese, String meat) {
        this.big = big;
        this.cheese = cheese;
        this.meat = meat;
    }

    public Boolean getBig() {
        return big;
    }

    public String getCheese() {
        return cheese;
    }

    public String getMeat() {
        return meat;
    }

    public Double getPrice() {
        double basePrice = Boolean.TRUE.equals(big) ? 2.0 : 1.0;
        double cheesePrice = cheese != null ? 1.0 : 0.0;
        double meatPrice = meat != null ? 1.0 : 0.0;
        return basePrice + cheesePrice + meatPrice;
    }
}

interface Builder<T> {

    Builder<T> big(boolean big);
    Builder<T> cheese(String cheese);
    Builder<T> meat(String meat);
    T getResult();
}

class PizzaBuilder implements Builder<Pizza> {

    private Boolean big;
    private String cheese;
    private String meat;

    @Override
    public PizzaBuilder big(boolean big) {
        this.big = big;
        return this;
    }

    @Override
    public PizzaBuilder cheese(String cheese) {
        this.cheese = cheese;
        return this;
    }

    @Override
    public PizzaBuilder meat(String meat) {
        this.meat = meat;
        return this;
    }

    @Override
    public Pizza getResult() {
        return new Pizza(this.big, this.cheese, this.meat);
    }
}

class ReceiptBuilder implements Builder<Receipt> {

    private Boolean big;
    private String cheese;
    private String meat;

    @Override
    public ReceiptBuilder big(boolean big) {
        this.big = big;
        return this;
    }

    @Override
    public ReceiptBuilder cheese(String cheese) {
        this.cheese = cheese;
        return this;
    }

    @Override
    public ReceiptBuilder meat(String meat) {
        this.meat = meat;
        return this;
    }

    @Override
    public Receipt getResult() {
        return new Receipt(this.big, this.cheese, this.meat);
    }
}

class Director {

    public <T> T constructBigNoCheesePizza(Builder<T> builder) {
        return builder
            .big(true)
            .meat("ham")
            .getResult();
    }

    public <T> T constructSmallNoMeatPizza(Builder<T> builder) {
        return builder
            .big(false)
            .cheese("mozzarella")
            .getResult();
    }
}

public class BuilderDemo {

    public static void main(String[] args) {
        Director director = new Director();
        Pizza bigNoCheesePizza = director.constructBigNoCheesePizza(new PizzaBuilder());
        Receipt receiptForBigNoCheesePizza = director.constructBigNoCheesePizza(new ReceiptBuilder());
        System.out.printf("Price is %s for %s%n", receiptForBigNoCheesePizza.getPrice(), bigNoCheesePizza.toString());
        Pizza smallNoMeatPizza = director.constructSmallNoMeatPizza(new PizzaBuilder());
        Receipt receiptForSmallNoMeatPizza = director.constructSmallNoMeatPizza(new ReceiptBuilder());
        System.out.printf("Price is %s for %s%n", receiptForSmallNoMeatPizza.getPrice(), smallNoMeatPizza.toString());
    }
}

File: design_patterns/builder/builder_python/src/builder_demo.py
from enum import Enum


class Pizza:

    def __init__(self, big: bool, cheese: str, meat: str):
        self.big = big
        self.cheese = cheese
        self.meat = meat

    def __str__(self):
        return "Pizza big=%s, " \
               "cheese=%s, " \
               "meat=%s" \
               % (self.big, self.cheese, self.meat)


class Receipt:

    def __init__(self, big: bool, cheese: str, meat: str):
        self.big = big
        self.cheese = cheese
        self.meat = meat

    def get_price(self) -> float:
        base_price = 2 if self.big is True else 1
        cheese_price = 0 if self.cheese is None else 1
        meat_price = 0 if self.meat is None else 1
        return base_price + cheese_price + meat_price


class Builder:

    def set_big(self, big: bool):
        pass

    def set_cheese(self, cheese: str):
        pass

    def set_meat(self, meat: str):
        pass


class PizzaBuilder(Builder):

    def __init__(self):
        self.big = None
        self.cheese = None
        self.meat = None

    def set_big(self, big: bool):
        self.big = big

    def set_cheese(self, cheese: str):
        self.cheese = cheese

    def set_meat(self, meat: str):
        self.meat = meat

    def get_result(self) -> Pizza:
        return Pizza(self.big, self.cheese, self.meat)


class ReceiptBuilder(Builder):

    def __init__(self):
        self.big = None
        self.cheese = None
        self.meat = None

    def set_big(self, big: bool):
        self.big = big

    def set_cheese(self, cheese: str):
        self.cheese = cheese

    def set_meat(self, meat: str):
        self.meat = meat

    def get_result(self) -> Receipt:
        return Receipt(self.big, self.cheese, self.meat)


class PizzaType(Enum):
    BIG_NO_CHEESE = 1
    SMALL_NO_MEAT = 2


class Director:

    def construct(self, pizza_type: PizzaType, builder: Builder):
        if pizza_type == PizzaType.BIG_NO_CHEESE:
            builder.set_big(True)
            builder.set_meat('ham')
        elif pizza_type == PizzaType.SMALL_NO_MEAT:
            builder.set_big(False)
            builder.set_cheese('mozzarella')


if __name__ == '__main__':
    director = Director()

    pizza_builder = PizzaBuilder()
    director.construct(PizzaType.BIG_NO_CHEESE, pizza_builder)
    big_no_cheese_pizza = pizza_builder.get_result()
    receipt_builder = ReceiptBuilder()
    director.construct(PizzaType.BIG_NO_CHEESE, receipt_builder)
    receipt_for_big_no_cheese_pizza = receipt_builder.get_result()
    print("Price is %s for %s" % (receipt_for_big_no_cheese_pizza.get_price(), big_no_cheese_pizza))

    pizza_builder = PizzaBuilder()
    director.construct(PizzaType.SMALL_NO_MEAT, pizza_builder)
    small_no_meat_pizza = pizza_builder.get_result()
    receipt_builder = ReceiptBuilder()
    director.construct(PizzaType.SMALL_NO_MEAT, receipt_builder)
    receipt_for_small_no_meat_pizza = receipt_builder.get_result()
    print("Price is %s for %s" % (receipt_for_small_no_meat_pizza.get_price(), small_no_meat_pizza))
